### ISSUE TITLE
WKTWriter precision

### ISSUE THREAD

**Issue by mukoki** [locationtech/jts#243]
Currently, WKTWriter writes coordinates using DecimalFormat with a limited precision of 1E-16 for doubles and 1E-6 for floats.
As a consequence writing geometries with WKTWriter and reading them back with WKTReader does not always preserve the geometry.

I suppose that the rational behind using DecimalFormat to write WKT is to avoid scientific notation (uncommon to write coordinates) but I did not find anything precise about it in OGC specification.

If we want to avoid scientific notation, I suggest to extend the number of decimal from 16 to 22 for doubles (which preserves full precision of coordinates as small as 1E-6, eventually useful for lat/lon coordinates) and from 6 to 10 decimals for floats.
I have a patch + a test for that but I want to make sure this is the right decision.

Another solution is to use standard java way of writing doubles, which always preserve the full precision of a double but which represents numbers like  0.0001 or 10 000 000 with scientific notation.

A mixed solution is to use decimal notation with full precision for numbers as large as 10 000 000 and as small as 0.000001 (using 22 decimals as proposed in the first solution) and switching to scientific notation for uncommon use cases using larger (resp. smaller) coordinates

**Comment by mukoki** [comment]
Sometimes, an example is better than a long talk. Here is a unit test which demonstrates the problem : 

```
public void testWriteNumberNeeding17decimals() throws ParseException{
    PrecisionModel precisionModel = new PrecisionModel(PrecisionModel.FLOATING);
    GeometryFactory geometryFactory = new GeometryFactory(precisionModel, 0);
    Point point1 = geometryFactory.createPoint(new Coordinate(0.39452545497198344d, 0));
    Geometry point2 = new WKTReader().read(new WKTWriter().write(point1));
    assertEquals(point1,point2);
  }
```

**Comment by dr-jts** [comment]
Yes, the rationale behind using DecimalFormat is to avoid scientific notation.   Agreed that this is not in the spec, but it seems like a good idea, since some parsers might not accept scientific notation. 

I was under the impression that IEEE DP supported 16 decimal digits of precision, which is where the 16 comes from.  But it seems Java at least provides at least 17 digits?  Is there a reference which documents this?

22 decimals seems like plenty!  And I suppose there is no harm in extending, since 16 is already a lot.  I wonder what this will do when converting a number which has a repeating decimal (as can happen sometimes - not sure it's easy to reproduce though).  

Does this mean that the PrecisionModel.getMaximumSignificantDigits needs to be changed too?  It is used elsewhere as well.

**Comment by dr-jts** [comment]
Really the output precision should be an option to WKTWriter - probably by allowing a PrecisionModel to be supplied independently of the Geometry.

**Comment by mukoki** [comment]
Hi,

I read somewhere that the precision of float/double is ~7.2/~15.9 decimals
But this good article explains why 9/17 digits are effectively necessary 
[https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html#1251]

In our case, I think replacing 6/16 digits in PrecisionModel by 9/17 would
be a first step to take this recommandation into account.

Using more decimal is generally not necessary for projected coordinates (a significant part of the precision is used by the integer part, so no need to add extra decimals to get the full double precision).
For geographic coordinates, I think that extra precision is useful to represent numbers less than 0.1 :
17 decimal digits are enough to represent, say 
0.39452545497198344
but we need 3 more digits if we want to represent exactly 
0.39452545497198344 E-3 -> 0.00039452545497198344

Note that DecimalFormat does not write trailing zeros, and I think in most cases (projected coordinates) adding precision in DecimalFormat definition  will not change the way coordinates are written.

I tested the change from 16 to 17 digits (in PrecisionModel.getMaximumSignificantDigits). It solves the case for 0.39452545497198344, and it only broke the test of PrecisionModel (which tests that precision is 16 ;-)

IMHO, it is not necessary to add an option in WKTWriter (who knows how many digits are necessary to preserve the full precision of a double ?). I think having a PrecisionModel which makes it possible to deal with Single precision or Double Precision is great (I never used Single precision) and a WKTWriter which preserves the precision of the PrecisionModel is the best way to respect the "Principle of least astonishment".

**Comment by dr-jts** [comment]
Makes good sense to raise the number of digits to 17, as per the classic Goldberg paper.

I'm not so sure that geodetic coordinates need that much precision.  1 degree is about 10^5 metres, so 17 decimal places is 10^-12 metres - which seems like it is far beyond the accuracy of any real geodetic data.

As for adding precision specification to WKTWriter, the reason for this is to support the case when the precision of the data is known, but not captured in the Geometry's precision model.  In theory the Geometry PrecisionModel should be used to express this, but in practice it's awkward to use different precision models in systems which provide more generic Geometry handling.  It's more convenient to be able to specify the precision at the point where it is directly needed.

**Comment by mukoki** [comment]
Sure, doubles generally have more precision than what we need. My concern is only about consistency.
If WKB -> WKT -> WKB transformation do not preserve original geometries, it means that using WKT in processes involving edit operation may break topology.
Even with 17 decimal digits, I think that the current statement of the API "The WKTWriter outputs coordinates rounded to the precision model" will not be 100% correct and the user should be warned.

I agree that it maybe more "convenient" to handle precision in WKT than in PrecisionModel. But when we know that rounding coordinates is enough to make a valid geometry become invalid, I'm not sure such an option is on the safe side.

**Comment by airbreather** [comment]
+1 to output ordinate values with their full precision when the user asks for a floating precision model.  I'm not quite sure the true purpose is of `FIXED`-type precision models, so I won't comment on them.

To my understanding, the basic principle behind floating-point is that people want to be able to discuss values in the neighborhood of 1.234E+30 the same way that they discuss values in the neighborhood of 1.234E+1 and 1.234E-30, yet [this code in `WKTWriter`](https://github.com/locationtech/jts/blob/a579291fbd1c1313652c701aa32e0d35e02e0030/modules/core/src/main/java/org/locationtech/jts/io/WKTWriter.java#L152-L162) appears to trim all the significant digits off of values near 1.234E-30 when using a floating precision model.

I'm sure this is fine for degrees and radians, but `WKTWriter` seems like an awkward place to rely on an assumption about the scale of the inputs, especially since I think the rest of the system handles these just fine.  It leads to situations like this in the test builder, where I type in "`POINT (1E-20 1E-20)`" and have it spit back out "POINT (0 0)" but still plot the point where I asked it:
![image](https://user-images.githubusercontent.com/3792815/53681859-4d026980-3cbd-11e9-9aab-6d158c3ef5e5.png)

Why not `"0." + stringOfChar('#', 324)` when it's floating-precision (cache globally to avoid building it every time)?  This is the minimum needed to print the [smallest value that's greater than zero](http://binaryconvert.com/result_double.html?hexadecimal=0000000000000001) in a way that unambiguously distinguishes it from both zero and [the second smallest value that's greater than zero](http://binaryconvert.com/result_double.html?hexadecimal=0000000000000002).

err... [it's that way in .NET](https://sharplab.io/#v2:EYLgHgbALANALiATgVwHYB8ACAGABJgRgG4BYAKB3wIioDoBhAe1QGdGAbAU1LPMwCZc9cgG9yuCVRqYouALIBDAJaoAFAEpxksWUl7cAE0bJgXXCwC2C9lxZxcAXlwAhJXCaoAbp0RwftAElUOGhXOBYAFUYAEWNTTlUCdR59SSMTM2sLRjsAZSsbTjtHFzcPb19/IJCoMMiYuK5VfmTyLVTCPAAzC3snZ2QldgMAMUZEKzhcuEQVAHMNHnb9dPjzAts4ACUIktWuWgAFBUQWBMtrTdoo6dnUBZ64dVbdVMNGzlwsnKmNou3dk59pwjiczqpvnk/nZroxbvNVI9nktXqkAOqzPwAGRUCQAJAAiC6FOwgN7kinaYmbAC+BJe5Ixbk4ONQ+KJ0IBZMpPKpnJ2dIZbyZ2NxqkJkN+l3+3N5bxEkvy0rsgpSwsxLLFEvY2Shyq5ctSCp1PyVJIBquWehFmrZ4o5+twAB4via9ebZfLqf9na7dVLzZbUfobaz2XgXd7SYaKSII+t9UHGRqw/ao30nOmdp7Y+nHJn+REk+rmantf6zZt837TYWycaK5zq4rC8WJDS2sHCNICHgBkNRuNJvD7hpljpyZ1zCU2QB3KjYVQAcgAxEuYLgABxJNUdADsuAJ2FoBNwAGppxeWOfL+Zd7gO2RH0A===), anyway.  I suspect it's the same in Java as well.

**Comment by dr-jts** [comment]
@airbreather In fact very small ordinate values like 1e-20 are unlikely to be well-handled by JTS operations, due to round-off error.  At least, they won't work if used with much larger ordinates as well.  They *might* work in ops if all ordinates have similar order of magnitude.  But are there really use cases for that?

**Comment by dr-jts** [comment]
This is an interesting discussion, and It certainly seems like JTS number output can be improved.  I think it would be nice to limit the number of sig figs output to something reflecting the precision of floating point - so maybe 17 or 18?  And for very small/large numbers use sci notation seems fine - especially since these are unlikely to occur very often in practice.

A PR for this is welcome. Might be a while until I get to it.

**Comment by dr-jts** [comment]
The whole issue of precision handling is quite tricky.  And trying to deal with it at the level of base-10 textual representations just adds to the complexity.  If the goal is to fully preserve precision then WKB should be used.  WKT should be seen as just a convenient human-readable representation.

It seems reasonable to increase the number of sig digits specified for the two floating precision models.  Some limited testing seems to indicate this doesn't cause any problems.  But it's really not a panacea for the overall problem.

**Comment by mukoki** [comment]
We are not far from my original proposition. I'll check if I have still a PR at hand.

Keeping the whole precision of a geometry after converting it forth and back from WKB to WKT seems an interesting goal and not so far from where we are. Indeed, Double.toString guarantee this with only a few more digits than WKTWriter, and also avoid scientific notation for a range as large as 1E-3 ; 1E8.
I think we just should avoid scientific notation for a range a bit larger
- indeed we can write numbers greater than 1E8 in decimal format at no coast. I guess 1E9 is enough for all geomatic purpose. Greater numbers may be more readable in scientific notation 
- For small numbers, decimal notation has a coast. If we want to keep full binary precision for a number with magnitude 1E-6, we have to add 6 extra zero. But coordinates under 1E-6 are (very) rare, so adding 6 digits may not be a problem. Below 1E-6 (or maybe 1E-9) we still can switch to scientific notation to be perfectly consistent with the binary precision.

**Comment by dr-jts** [comment]
Is it really possible to convert lossless between numeric and binary FP?  Aren't there issues with numbers that can't be represented exactly in FP?

If it is possible, then it seems like a good goal.

Agreed that we should only use sci not for extremely large or small numbers.

**Comment by airbreather** [comment]
> Is it really possible to convert lossless between numeric and binary FP? Aren't there issues with numbers that can't be represented exactly in FP?

Just responding to this quote right now...

Indeed, "0.3" cannot be represented exactly.  In IEEE-754 double-precision binary floating-point:
- 0.299999999999999988897769753748 is the closest you can get.
- 0.299999999999999933386618522491 is the next-smallest.
- 0.300000000000000044408920985006 is the next-largest.

The trick is that we don't need to have a value that's exactly equal to "0.3" in order to round-trip.  If we have the value from the first bullet point, then we can safely represent it in text as "0.3".  There is no legal value that's closer to 0.3, so there's no risk of it getting interpreted as something else.

If we have either of the other two values, however, then the only way to guarantee round-trip to/from string is to write at least 17 digits to the right of the decimal point:
- For the next-smallest, stopping at 16 gives you a value closer to 0.299999999999999877875467291233
- For the next-largest, stopping at 16 gets you right back down to 0.3

However, we are allowed to stop at 17 of the digits to the right of the decimal point (not the full 30) because there's no ambiguity that the extra digits would help us resolve... except what happens with 0.03?
- 0.0299999999999999988897769753748 is the closest you can get.
- 0.0299999999999999954203300234212 is the next-smallest.
- 0.0300000000000000023592239273285 is the next-largest.

Here, you need to write 18 digits to the right of the decimal point in order to get either of the bottom two values to round-trip, even though 17 digits were plenty for values near 0.3.  So if we have a format string of `0.###...###` (with 17 `#` characters), then our output will not be able to represent these values.

I haven't done the full detailed deep-dive for values near 0.003, but I expect the same kind of thing: you need another digit to the right of the decimal point to avoid losing precision.

Normally, the system is supposed to handle all of this for us; indeed, I'm pretty sure that Java will work just fine if we don't try to do anything special / fancy when given a floating `PrecisionModel` (provided that we're OK with scientific notation beyond a point).  This makes sense to me, because keeping all the precision that's stored on the value (for better or for worse) is what I thought the idea was behind floating `PrecisionModel` instances in the first place :grin:.

**Comment by dr-jts** [comment]
@mukoki This issue is mostly resolved by PR #409, I think - would you like to review?

I guess one thing that can be done is to bump up the default maximum fraction digits to 22 as per your recommendation - I'm fine with doing this.

**Comment by mukoki** [comment]
Good news!
Yes, I think 22 decimals places covers most cases in geospatial applications either in degrees or in meters, without loosing the full double precision. It is somewhat arbitrary though. For big numbers, I think there is no problem, but for numbers starting with 0.000..., the more leading 0 there are, the more precision we need if we want to keep the full precision of the floating point.
Anyway, if I am right, adding decimal places in OrdinateFormat should not add too much overhead as most of times it should add trailing 0 which must be omitted

I did not find time to test your code yet. I have two small observations after reading it : 
OrdinateFormat
Seems that the OrdinateFormat() constructor could be private as it seems
irrelevant to build several formatter with the same default MAX_FRACTION_DIGITS

WKTWriter
The PrecisionModel is defined at the WKTWriter level,
and only if null, WKTWriter uses the PrecisionModel of the geometry.
On the other hand, a new OrdinateFormat is created for each Write 
operation (even if it uses the WKTWriter PrecisionModel).
I think it could be cached in the WKTWriter for the former case,
or maybe in the PrecisionModel for both cases.

**Comment by dr-jts** [comment]
> Yes, I think 22 decimals places covers most cases in geospatial applications either in degrees or in meters, without loosing the full double precision. It is somewhat arbitrary though. For big numbers, I think there is no problem, but for numbers starting with 0.000..., the more leading 0 there are, the more precision we need if we want to keep the full precision of the floating point.
> Anyway, if I am right, adding decimal places in OrdinateFormat should not add too much overhead as most of times it should add trailing 0 which must be omitted

Agreed that numbers like 0.0000...  will potentially drop digits.  Perhaps @airbreather 's suggestion about using 340 would solve this.  It seems like overkill, but perhaps is unnoticeable except when needed.  I will experiment.

**Comment by dr-jts** [comment]
> OrdinateFormat
> Seems that the OrdinateFormat() constructor could be private as it seems
> irrelevant to build several formatter with the same default MAX_FRACTION_DIGITS

Could be.  Even if the constructors are private new instances can still be created via the public `create` methods. I didn't do this because that's not been the pattern in JTS so far (for no very good reason).

> WKTWriter
> The PrecisionModel is defined at the WKTWriter level,
> and only if null, WKTWriter uses the PrecisionModel of the geometry.
> On the other hand, a new OrdinateFormat is created for each Write
> operation (even if it uses the WKTWriter PrecisionModel).
> I think it could be cached in the WKTWriter for the former case,
> or maybe in the PrecisionModel for both cases.

Agreed, technically speaking.  Using the geometry precision model I think is by far the most common use case, so caching the OrdinateFormat when the PM is supplied explicitly probably doesn't save that many cycles.  But that is fairly easy to do, so will make that change.

**Comment by dr-jts** [comment]
It looks like using a maxFractionDigits of 340 doesn't cause any harm, so will switch to that.
